modiscolite/io.py

```diff
@@ -158,6 +163,59 @@ def save_hdf5(filename, pos_patterns, neg_patterns):
 			save_pattern(pattern, neg_pattern)
 
 
+def write_meme_from_h5(filename: os.PathLike, datatype: util.MemeDataType, output_filename: os.PathLike) -> None:
+	"""Write a MEME file from an h5 file output from TF-MoDISco. Based on the given datatype.
+
+	Parameters
+	----------
+	filename: str
+		The name of the h5 file to read.
+	datatype: MemeDataType
+		The datatype to use for the MEME file.
+	output_filename: str
+		The name of the MEME file to write.
+	"""
+
+	writer = meme_writer.MemeWriter(
+		memesuite_version='5',
+		alphabet='ACGT',
+		background_frequencies='A 0.25 C 0.25 G 0.25 T 0.25'
+	)
+
+	with h5py.File(filename, 'r') as grp:
+		for (name, datasets) in grp['pos_patterns'].items():
+
+			probability_matrix = None
+			if datatype == util.MemeDataType.PFM:
+				probability_matrix = datasets['sequence'][:] / np.sum(datasets['sequence'][:], axis=1, keepdims=True)
+			elif datatype == util.MemeDataType.CWM:
+				probability_matrix = datasets['contrib_scores'][:]
+			elif datatype == util.MemeDataType.hCWM:
+				probability_matrix = datasets['hypothetical_contribs'][:]
+			elif datatype == util.MemeDataType.CWM_PFM:
+				# Softmax version of CWM.
+				probability_matrix = np.exp(datasets['contrib_scores'][:]) / np.sum(np.exp(datasets['contrib_scores'][:]), axis=1, keepdims=True)
```

Comment:
Use https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.softmax.html instead. Mathematically you're correct, but there are numerical stability issues that can arise if you do it this way.

Reply:
Done. Confirmed that they provide same values with my own example